plugins/scaffolder-backend/src/scaffolder/actions/builtin/publish/githubPullRequest.ts

```diff
@@ -0,0 +1,231 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import fs from 'fs';
+import path from 'path';
+import { promisify } from 'util';
+
+import {
+  GithubCredentialsProvider,
+  ScmIntegrationRegistry,
+} from '@backstage/integration';
+import { zipObject } from 'lodash';
+import { createTemplateAction } from '../../createTemplateAction';
+import { Octokit } from '@octokit/rest';
+import { OctokitResponse } from '@octokit/types';
+import { InputError } from '@backstage/errors';
+import { createPullRequest } from 'octokit-plugin-create-pull-request';
+import globby from 'globby';
+
+const readFile = promisify(fs.readFile);
+
+type PullRequestResponse = {
+  html_url: string;
+};
+
+export interface PullRequestCreator {
+  createPullRequest(
+    options: createPullRequest.Options,
+  ): Promise<OctokitResponse<PullRequestResponse> | null>;
+}
+
+export type GithubPullRequestActionInput = {
+  owner: string;
+  repo: string;
+  title: string;
+  branchName: string;
+  description: string;
+  host?: string;
+  targetPath?: string;
+  sourcePath?: string;
+};
+
+export type ClientFactoryInput = {
+  host: string;
+  owner: string;
+  repo: string;
+};
+
+export const createClientFactory = async (
+  integrations: ScmIntegrationRegistry,
+  { owner, repo, host = 'github.com' }: ClientFactoryInput,
+): Promise<Octokit> => {
+  const integrationConfig = integrations.github
+    .list()
+    .find(integration => integration.config.host === host)?.config;
+
+  if (!integrationConfig) {
+    throw new InputError(`No integration for host ${host}`);
+  }
+
+  const credentialsProvider = GithubCredentialsProvider.create(
+    integrationConfig,
+  );
+
+  if (!credentialsProvider) {
+    throw new InputError(
+      `No matching credentials for host ${host}, please check your integrations config`,
+    );
+  }
+
+  const { token } = await credentialsProvider.getCredentials({
+    url: `${host}/${encodeURIComponent(owner)}/${encodeURIComponent(repo)}`,
+  });
+
+  if (!token) {
+    throw new InputError(
+      `No token available for host: ${host}, with owner ${owner}, and repo ${repo}`,
+    );
+  }
+
+  return new Octokit({
+    auth: token,
+    baseUrl: integrationConfig.apiBaseUrl,
+  });
+};
+export const defaultClientFactory = async (
+  integrations: ScmIntegrationRegistry,
+  { owner, repo, host = 'github.com' }: GithubPullRequestActionInput,
+): Promise<PullRequestCreator> => {
+  const baseClient = await createClientFactory(integrations, {
+    owner,
+    repo,
+    host,
+  });
+  const client = createPullRequest(baseClient);
+
+  return client;
+};
+
+interface CreateGithubPullRequestActionOptions {
+  integrations: ScmIntegrationRegistry;
+  clientFactory?: (
+    integrations: ScmIntegrationRegistry,
```

Comment:
Best make this first argument an object in case we want to provide more things in the future.

Reply:
The params of the `clientFactory`, not the action create function :p

Reply:
Ah, I see. I smooshed them both into one options param.